### PR TITLE
Add a failing test case with `types: null` condition

### PR DESCRIPTION
**Pull request by Andarist**
This is a test case related to this bug: https://github.com/microsoft/TypeScript/issues/50762 .

It feels that this situation is specific-enough to be fixable regardless of what is decided for #50762 , cc @andrewbranch

### CODE REVIEW

File: tests/baselines/reference/disallowThroughTypesConditionAndAvoidFallback(moduleresolution=nodenext).errors.txt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+error TS6504: File '/node_modules/foo/index.js' is a JavaScript file. Did you mean to enable the 'allowJs' option?
```
Comment (Andarist) [inline tests/baselines/reference/disallowThroughTypesConditionAndAvoidFallback(moduleresolution=nodenext).errors.txt:1]: i'm not sure how to avoid this in the test setup without `allowJs: true` (which I would prefer to avoid here).

File: tests/baselines/reference/disallowThroughTypesConditionAndAvoidFallback(moduleresolution=nodenext).errors.txt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    
+==== /node_modules/foo/index.js (0 errors) ====
+    module.exports = { a: 'a' };
+    
+==== /node_modules/foo/index.d.ts (0 errors) ====
+    export declare const a: 'a'
+    
+==== /a.ts (0 errors) ====
+    export { a } from "foo";
```
Comment (Andarist) [inline tests/baselines/reference/disallowThroughTypesConditionAndAvoidFallback(moduleresolution=nodenext).errors.txt:26]: I feel like this should error because of the quite explicit `types: null` declaration in the package

**Comment by andrewbranch** [comment]
I agree that this is #50762 which makes it a bug, and I agree that it’s narrow enough to be fixable on its own, and at first glance at the conversation at https://github.com/discordjs/discord.js/pull/9485 it even feels useful. But thinking about it a bit more, it feels a little suspicious to try to forbid usage of a package subpath by derailing TypeScript specifically. In my view, TypeScript’s job is to know how the runtime/bundler is going to resolve, and then to find types for the implementation file the runtime/bundler finds. Sometimes, though, you need to sacrifice a bit of the former in service of the latter—that is, if your types aren’t colocated with implementation files, you grab TypeScript with the `"types"` condition earlier in the process and say “look, use this file over here, and just _trust_ me that it’s a representation of what the runtime/bundler will resolve to.” But the usage you suggest in the discord.js conversation would make the `"types"` condition a lie. I think it would be better for them to use other nested conditions, or move the subpath to `"imports"`, or whatever, in order to make that subpath effectively private to consumers, which seems to be the effect they’re going for.

**Comment by Andarist** [comment]
Yeah, I overall agree with you - it's fishy to disallow something by disallowing `types` specifically. I don't know enough about their use case to suggest them a better way though. It just came up there since they *thought* that by not including `types` condition for this entrypoint they would forbid that at type-level.

However, because of the "used fallback condition" bug it didn't actually end up being forbidden there and even its more explicit variant (`types: null`) doesn't do that. 

This particular PR is more about the fact that this should work and not as much about if it should be used 😉

**Comment by Andarist** [comment]
It's also very likely that this particular manifestation of this bug makes weird combinations to behave incorrectly, like `import: null, require: './index.js'`. Or perhaps even stuff like `node: null, import: './index.mjs'`.

**Comment by andrewbranch** [comment]
Do you want to see what the narrow fix for this would look like? If we don’t have to jump through hoops to fix `null` without fixing the rest of #50762, I’m not against it.

**Comment by Andarist** [comment]
> Do you want to see what the narrow fix for this would look like? 

Do I want to? Not really. Will I investigate this? Yeeeah, that's very likely.

**Comment by andrewbranch** [comment]
I’m going to steal this PR and fix it

**Comment by Andarist** [comment]
Be my guest, take It 😉

**Comment by andrewbranch** [comment]
Suppose you have

```json5
// @Filename: /a/b/c/node_modules/foo/package.json
{
  "exports": {
    ".": {
      "types": null
    }
  }
}
```

and

```json5
// @Filename: /a/b/c/node_modules/@types/foo/package.json
{
  "exports": {
    ".": {
      "types": "./index.d.ts"
    }
  }
}
```

Should the former block the latter? Should the former block a lookup in a higher-up node_modules directory?

**Comment by Andarist** [comment]
Yes. The first matched location like this should stop the algorithm from looking any further. 

I’d probably extend this to just any condition (not just to `types`). If `null` value gets selected that is a pretty explicit sign that the matched entrypoint is forbidden/disallowed at runtime (and thus at types level too)

**Comment by andrewbranch** [comment]
I think this has some fairly unintuitive semantics. `"types": null` looks like it might mean “explicitly not typed,” but as you said, the fact that we arrived at `null` through `"types"` specifically is immaterial. If TS resolution arrives at a `null`, it should indicate that runtime resolution will also arrive at a `null` and stop further lookups.

If you want to tell TypeScript that something is explicitly not typed... you can point `"types"` at an untyped JavaScript file? 😅 But that _won’t_ actually stop us from looking elsewhere for types.

I think this interpretation is perfectly consistent... it just struck me as weird. Implementation-wise, we actually have _no existing mechanism_ to abort resolution with a failure. Sigh.

**Comment by jaridmargolin** [comment]
@andrewbranch - Are you still attempting to fix this? If I understand this thread correctly, I have what I would consider a similarly narrow issue. I am attempting to "hide" a subset of modules when using a custom "browser" condition. Would it make sense to create a new issue?

**lib package.json**:
```
  "exports": {
    ...
    "./entities/*/node": {
      "browser": null,
      "node": "./dist/entities/*/index.node.js"
    },
  }
```

**consuming browser app tsconfig.json**:
```
  "compilerOptions": {
    "customConditions": ["browser"]
  }
```

**Comment by andrewbranch** [comment]
Planning to fix this as a breaking change in 6.0 or 6.5, depending on whether it needs a deprecation message; I’m not sure we have the details settled.

**Comment by andrewbranch** [comment]
I'm pretty sure this will fail CI because the behavior changed for the better in https://github.com/microsoft/TypeScript/pull/62483, which includes relevant tests. I just wanted to see before closing it permanently.
